### PR TITLE
[Breaking] Replace DART with tree subsampling

### LINKED ISSUE #12339: [RFC] Introduce a Single "dropout" Parameter for DART
**Summary**

I propose introducing a new DART parameter:

dropout ∈ [0, 1]

and deprecating the existing DART-specific parameters:

- "rate_drop"
- "sample_type"
- "normalize_type"
- "one_drop"
- "skip_drop"

The new "dropout" parameter would implement the canonical DART algorithm described in the original paper with the following fixed behavior:

- "sample_type = uniform"
- "normalize_type = tree"
- "one_drop = false"
- "skip_drop = 0"

The existing parameters can remain supported during a deprecation period for backwards compatibility.

**Motivation**

The original DART paper defines a remarkably simple algorithm:

1. Drop each previous tree independently with probability "dropout".
2. If no tree is selected, drop one uniformly chosen tree ("Binomial-plus-one").
3. Train the new tree.
4. Apply a fixed normalization scheme.

The paper does not experimentally justify alternative sampling schemes, normalization methods, or skipping dropout iterations. These options were introduced later as implementation extensions rather than core components of DART.

Today XGBoost exposes five DART-specific parameters, making DART considerably more complicated than both the original algorithm and the analogous dropout mechanism used in neural networks, where users typically tune a single probability.

**Rationale**

"dropout"

This is the fundamental DART hyperparameter and should be the only user-facing regularization parameter.

"one_drop"

The original paper always ensures that at least one tree is dropped. This is a boundary condition rather than an independent modeling choice.

Without this rule, for small dropout probabilities,

$P(K=0)=(1-p)^T$,

meaning many boosting rounds perform no dropout at all and degenerate to ordinary GBDT.

I would propose that the user should increase the dropout probability instead of using one drop.

"sample_type"

The paper uses uniform sampling and does not provide empirical evidence comparing it with weighted sampling.

Uniform sampling should therefore become the canonical behavior.

"normalize_type"

The paper defines a single normalization rule.

While XGBoost currently supports multiple normalization strategies, there is little published evidence that exposing this choice as a user-facing hyperparameter provides consistent benefit.

"skip_drop"

The original algorithm performs dropout every boosting iteration.

"skip_drop" effectively introduces a second regularization parameter controlling dropout frequency in addition to dropout probability. This increases tuning complexity without clear evidence that users benefit from controlling these effects independently.

**Benefits**

- A significantly simpler and more intuitive API.
- Closer alignment with the original DART algorithm.
- Easier parameter tuning.
- Consistency with dropout interfaces in deep learning, where a single probability controls regularization.
- Existing models remain compatible during a deprecation period while new users are presented with a much simpler interface.

### PR DESCRIPTION
**Pull request by RAMitchell**
Closes https://github.com/dmlc/xgboost/issues/12339.

## Summary

Replace DART tree-reweighting normalization with expectation-preserving tree subsampling applied to the temporary training prediction.

For tree retention probability $q$, each existing tree is independently retained with

```math
m_i \sim \mathrm{Bernoulli}(q),
```

and gradients are calculated from

```math
\widetilde{F}(x) = F_0(x) + \frac{1}{q}\sum_i m_i F_i(x).
```

The base score or base margin $F_0$ is protected from normalization, giving

```math
\mathbb{E}\left[\widetilde{F}(x)\right] = F(x).
```

The new tree is committed normally without modifying the weights of existing or newly added trees.

## Motivation

Existing DART normalization repeatedly modifies historical tree weights. This changes the effective convergence behavior and complicates prediction caching, serialization, model continuation, and inference.

The new formulation treats the existing trees as a sampled training input:

- Existing trees retain fixed coefficients.
- New trees are committed normally.
- Newly trained models do not create or serialize `weight_drop`.
- Inference uses the ordinary additive-tree path.
- Normalization is applied only to the temporary training prediction.

## Parameters and migration

- Add `tree_subsample`, the probability of retaining each existing tree, with $0.000001 \leq q \leq 1$ and default $q=1$.
- When `tree_subsample` is absent, deprecated `rate_drop=r` is converted to `tree_subsample=1-r`.
- The conversion warning explains that this preserves the uniform tree-retention probability but not legacy DART normalization, and links https://github.com/dmlc/xgboost/issues/12339.
- If both parameters are supplied, `tree_subsample` takes precedence.
- `sample_type`, `normalize_type`, `one_drop`, and `skip_drop` have no exact conversion. They are ignored with removal warnings that link the same issue.
- `booster=dart` remains accepted as a deprecated alias for `gbtree` and warns that legacy DART normalization has been removed.

## Relation to row subsampling

Tree subsampling introduces variance into training and can help reduce overfitting similarly to row subsampling.

Row subsampling perturbs only the tree fitted in the current round, so its relative influence on the total ensemble prediction diminishes as the ensemble grows. Tree subsampling perturbs every accumulated tree at the same per-tree probability, so its effect does not vanish merely because the ensemble contains more trees.

Consequently, the omitted-tree fraction `1 - tree_subsample` will often need to be smaller than the omitted-row fraction `1 - subsample`.

## Compatibility

This is an algorithmically breaking change: the legacy DART normalization algorithm is removed rather than emulated. Existing configurations remain parseable, and `rate_drop` receives an explicit conversion and behavioral warning instead of being silently ignored.

Legacy weighted DART models remain loadable and retain their historical tree weights for prediction. New rounds appended to legacy models are committed with unit weight.

## Performance

This PR does not attempt to optimize sampled training prediction. The current implementation expresses the random sample through temporary prediction weights and still traverses the full ensemble. A future optimization can predict only the omitted trees and subtract their contribution from the cached full prediction.

## Testing

- CPU library and `testxgboost` build successfully.
- Focused native parameter migration, serialization, unbiasedness, sampled-prediction, and deprecated-alias tests pass.
- Focused Python tree-subsampling and multi-output tests pass.
- Property-based serialized training-continuation testing passes.
- C++ formatting, C++ lint, and `git diff --check` pass.

CUDA, R, JVM, and Ruff validation are left to CI.

### CODE REVIEW

File: src/gbm/gbtree.h [hunk 1]
```diff
@@ -97,34 +100,38 @@ struct DartTrainParam : public XGBoostParameter<DartTrainParam> {
   float skip_drop;
 
   DMLC_DECLARE_PARAMETER(DartTrainParam) {
+    DMLC_DECLARE_FIELD(dropout_rate)
+        .set_range(0.0f, 1.0f - kRtEps)
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/gbm/gbtree.h:104]: Kept the deliberate numerical guard and documented the actual supported range as `[0.0, 0.999999]` in the parameter reference, tutorial, R help, Spark validation, and PR description. Fixed in 94488a06b.

File: src/gbm/gbtree.h [hunk 2]
```diff
@@ -97,34 +100,38 @@ struct DartTrainParam : public XGBoostParameter<DartTrainParam> {
...
         .set_range(0.0f, 1.0f)
         .set_default(0.0f)
-        .describe("Fraction of trees to drop during the dropout.");
+        .describe("Deprecated and ignored. Use dropout_rate instead.");
     DMLC_DECLARE_FIELD(one_drop).set_default(false).describe(
-        "Whether at least one tree should always be dropped during the dropout.");
+        "Deprecated and ignored. Independent dropout can retain all trees.");
     DMLC_DECLARE_FIELD(skip_drop)
-        .set_range(0.0f, 1.0f)
+        .set_range(0.0f, 1.0f - kRtEps)
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/gbm/gbtree.h:124]: Updated `skip_drop` documentation and Spark validation to the same `[0.0, 0.999999]` range as `dropout_rate`. Fixed in 94488a06b.

File: tests/python/test_multi_target.py [hunk 3]
```diff
@@ -233,9 +233,7 @@ def test_dart_normalization_multi_output_eta() -> None:
             "base_score": 0.0,
             "reg_lambda": 0.0,
             "min_child_weight": 0.0,
-            "rate_drop": 0.0,
-            "one_drop": True,
-            "normalize_type": "tree",
+            "dropout_rate": 0.0,
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply tests/python/test_multi_target.py]: Set `dropout_rate=0.5` and replaced the ordinary-boosting expectation with a mask-independent dropout assertion: each output is either 0 or 2 after two rounds. Fixed in 94488a06b.

**Review by trivialfis — COMMENTED** [review]
I suggest that we don't use `dropout` and replace it with `tree_subsample` or something similar, to keep the semantics of sampling consistent: Proba `p` of something being used, instead of something being dropped.

**Comment by RAMitchell** [comment]
I like that idea. The downside is there is no deprecation - DART gets suddenly removed.

**Comment by RAMitchell** [comment]
I ran a small experiment to characterize the proposed tree subsampling against ordinary XGBoost and legacy DART with both tree and forest normalization.

All methods use native XGBoost with matched settings: 600 rounds, learning rate 0.05, depth 4, and no row or column subsampling. Each dataset uses five paired 60/40 train/test splits. Tree subsampling and both DART variants omit each tree with probability 0.1 (`tree_subsample=0.9` for the new method).

### Training curves

image

The new method follows ordinary boosting closely, while both DART normalization schemes substantially slow training-loss reduction, particularly the default tree normalization. This supports separating tree sampling from historical tree reweighting.

### Held-out curves

image

Held-out performance is close to ordinary boosting: slightly better on `make_regression` and slightly worse on the other three datasets. Forest DART achieves slightly lower final held-out loss on Friedman1 despite its slower training-loss reduction.

Curves show mean full-ensemble loss, with bands indicating ±1 standard deviation across splits. Loss axes are logarithmic. California Housing uses a fixed 5,000-row subset; the other datasets are noisy Friedman1, noisy `make_regression`, and Breast Cancer Wisconsin.
